src/goto-symex/symex_goto.cpp

```diff
@@ -126,12 +146,62 @@ void goto_symext::symex_goto(statet &state)
     state_pc=goto_target;
   }
 
+  // Goto is conditional, so figure out which branch to take before
+  // checking whether to do path exploration or model-checking, and before
+  // adjusting guards.
+  std::string gp=options.get_option("goto-priority");
+  if(gp=="jump")
+  {
+    goto_programt::const_targett tmp=new_state_pc;
+    new_state_pc=state_pc;
+    state_pc=tmp;
+  }
+  else
+    INVARIANT(gp=="next",
+        "Unknown goto-priority option '" + gp + "'. Use one of "
+        "'jump' or 'next'.");
+
+  // Normally the next instruction to execute would be state_pc and we save
+  // new_state_pc for later. But if we're executing from a saved state, then
+  // new_state_pc should be the state that we saved from earlier, so let's
+  // execute that instead.
+  if(state.has_saved_target)
+  {
+    INVARIANT(new_state_pc==state.saved_target,
+        "Tried to explore the other path of a branch, but the next "
+        "instruction along that path is not the same as the instruction "
+        "that we saved at the branch point. Saved instruction is " +
+        state.saved_target->code.pretty() + "\nwe were intending "
+        "to explore " + new_state_pc->code.pretty() + "\nthe "
+        "instruction we think we saw on a previous path exploration is " +
+        state_pc->code.pretty());
+    goto_programt::const_targett tmp=new_state_pc;
+    new_state_pc=state_pc;
+    state_pc=tmp;
+    std::cerr << "Resuming from '" << state_pc->code.source_location() << "'\n";
```

Comment:
If this is a debug message, guard with `#ifdef DEBUG_GOTO_SYMEX` or something (and the same for `iostream` include above). Otherwise, use a `messaget` or get one passed in.

Reply:
If #1625 gets merged in before this, symext will have a messaget mutable member as you suggested so I'll rebase and use that

Reply:
FWIW I'll merge #1625 as soon as CI completes.